s2spy/traintest.py

```diff
@@ -2,52 +2,94 @@
-    consecutive folds.
-
-    For more information
+def fold_by_anchor(cv, data: Union[xr.Dataset, pd.DataFrame]):
+    """Splits calendar resampled data into train/test groups, based on the anchor year.
+    As splitter, a Splitter Class such as sklearn's KFold can be passed.
 
-    args:
-        df: DataFrame to store the train/test groups. It must have a column named "anchor_years".
-        n_splits: number of train/test splits.
+    For an overview of the sklearn Splitter Classes see:
+    https://scikit-learn.org/stable/modules/classes.html#module-sklearn.model_selection
 
-    Other keyword arguments: see the documentation for sklearn.model_selection.KFold:
-    https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html
+    Args:
+        cv (Splitter Class): Initialized splitter class, much have a `fit(X)` method
+            which splits up `X` into multiple folds of train/test data.
+        data (xr.Dataset or pd.DataFrame): Dataset with `anchor_year` as dimension or
+            DataFrame with `anchor_year` as column. E.g. data resampled using
+            `s2spy.time.AdventCalendar`'s `resample` method.
 
     Returns:
-        Pandas DataFrame with columns containing train/test label in each fold.
+        The input dataset with extra coordinates or columns added for each fold,
+            containing the labels 'train', 'test', and possibly 'skip'.
     """
-    cv = KFold(n_splits, shuffle=shuffle, **kwargs)
-    folds = cv.split(df)
+    assert isinstance(
+        data, (xr.Dataset, pd.DataFrame)
+    ), "Input data should be of type xr.Dataset or pd.DataFrame"
 
-    # Map folds to a new dataframe
-    output = pd.DataFrame(index=df.index)
-    for i, (train_index, test_index) in enumerate(folds):
-        col_name = f"fold_{i}"
-        output[col_name] = "skip"
-        output[col_name].iloc[train_index] = "train"
-        output[col_name].iloc[test_index] = "test"
+    if isinstance(data, xr.Dataset):
+        data = data.copy()  # otherwise the dataset is modified inplace
 
-    return output
+        assert (
+            data.anchor_year.size > 1
+        ), "Input data must have more than 1 anchor year to split."
 
+        folds = cv.split(data.anchor_year)
 
-def random_strat():
-    """random_strat cross-validator."""
-    raise NotImplementedError
+        fold_data = np.empty(
+            (cv.get_n_splits(data.anchor_year.size), data.anchor_year.size), dtype="<U6"
+        )
+
+        for i, (train_indices, test_indices) in enumerate(folds):
+            fold_data[i, :] = "skip"
+            fold_data[i, train_indices] = "train"
+            fold_data[i, test_indices] = "test"
+
+        data.expand_dims("fold")
+        data["fold"] = np.arange(fold_data.shape[0])
+        data["traintest"] = (["fold", "anchor_year"], fold_data)
+        data = data.set_coords("traintest")
+
+    else:
+        # split the anchor years
```

Comment:
```suggestion
    else:
        # input data is a pandas dataframe
```

Reply:
When I read this I am very tempted to split it into three functions:

```Py
def fold_by_anchor(cv, data: Union[xr.Dataset, pd.DataFrame]):
    """This is the public API so full docstring comes here."""
    if isinstance(data, xr.Dataset):
        return _fold_dataset(cv, data)
    elif isinstance(data, pd.DataFrame):
        return _fold_dataframe(cv, data)
    raise TypeError("Input data should be xr.dataset or pd.dataframe")


def _fold_dataset(cv, data: xr.Dataset):
   # implementation here


def _fold_dataframe(cv, data: pd.DataFrame):
   # implementation here

```

That also neatly gets rid of all your assertions.

I was trying to find where I picked this up, but I can't find a good source. Anyway, it relates to best practices such as "a function should do only one thing and do one thing well", "a function should not be too long", and the idea of multiple dispatch (one of the main idioms of the Julia language).